WalletWasabi.Tests/UnitTests/WabiSabi/Backend/Filters/IdempotencyActionFilterTests.cs

```diff
@@ -0,0 +1,73 @@
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.Caching.Memory;
+using Moq;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using WalletWasabi.Backend.Filters;
+using Xunit;
+
+namespace WalletWasabi.Tests.UnitTests.WabiSabi.Backend
+{
+	public class IdempotencyActionFilterTests
+	{
+		[Fact]
+		public void ShouldCacheResponse()
+		{
+			var body = Encoding.UTF8.GetBytes("Hello world");
+			using var stream1 = new MemoryStream(body);
+			var actionContext = CreateActionExecutingContext(stream1);
+			var cache = new MemoryCache(new MemoryCacheOptions());
+			var filter = new IdempotentActionFilterAttributeImpl(cache);
+
+			filter.OnActionExecuting(actionContext);
+
+			var resultContext = CreateResulExecutedContext(actionContext);
+			filter.OnResultExecuted(resultContext);
+
+			using var stream2 = new MemoryStream(body);
+			actionContext = CreateActionExecutingContext(stream2);
+			filter.OnActionExecuting(actionContext);
+
+			Assert.Equal(1, cache.Count);
+
+			var returned = Assert.IsType<OkObjectResult>(resultContext.Result);
+			var objValue = Assert.IsType<KeyValuePair<string, string>>(returned.Value);
+			Assert.Equal("Greeting", objValue.Key);
+			Assert.Equal("Hello world", objValue.Value);
+		}
```

Comment:
if i'm not mistaken this doesn't ensure the data was returned from the cache, only that data was entered into it under the same key, right? can the mock controller keep track of how many times the cached handler was invoked easily?

Reply:
You're right. I changed it to make clear the response doesn't comes from the controller but from somewhere else because it returns an _old_ value.